File: epi_judge_java/epi/test_framework/TestFailure.java

package epi.test_framework;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Tested function is allowed to throw this exception
 * in order to mark the current test as failed.
 */
public class TestFailure extends Exception {
  private String description;
  private List<Property> properties;

  public enum PropertyName {
    EXCEPTION_MESSAGE, // message of unhandled exception
    EXPLANATION,       // explanation from TSV file
    COMMAND,  // last command, that caused the error, in API-testing programs
    STATE,    // state of the user-defined collection (for instance, in API
              // testing)
    EXPECTED, // expected result
    RESULT,   // user-produced result
    MISSING_ITEMS,  // list of items from input that are missing in the result
                    // set
    EXCESS_ITEMS,   // list of items from result that are missing in the input
                    // set
    MISMATCH_INDEX, // for collections: index of the wrong item in result
                    // for binary trees: instance of TreePath describing the
                    // position of the wrong item
    EXPECTED_ITEM,  // value of the expected item in collection (not the whole
                    // collection)
    RESULT_ITEM     // value of the result item in collection (not the whole
                    // collection)
  }

  public class Property {
    private PropertyName name;
    private Object value;

    public Property(PropertyName name, Object value) {
      this.name = name;
      this.value = value;
    }

    public String name() {
      return name.toString().toLowerCase().replace('_', ' ');
    }

    public PropertyName rawName() { return name; }

    public int id() { return name.ordinal(); }

    public Object value() { return value; }
  }

  public TestFailure() {
    properties = new ArrayList<>();
    description = "";
  }

  public TestFailure(String description) {
    properties = new ArrayList<>();
    this.description = description;
  }

  public TestFailure withProperty(PropertyName name, Object value) {
    properties.add(new Property(name, value));
    return this;
  }

  public TestFailure withMismatchInfo(Object mismatchIndex, Object expectedItem,
                                      Object resultItem) {
    return withProperty(PropertyName.MISMATCH_INDEX, mismatchIndex)
        .withProperty(PropertyName.EXPECTED_ITEM, expectedItem)
        .withProperty(PropertyName.RESULT_ITEM, resultItem);
  }

  public String getDescription() { return description; }

  public int getMaxPropertyNameLength() {
    return properties.stream()
        .mapToInt(property -> property.name().length())
        .max()
        .orElse(0);
  }

  public List<Property> getProperties() {
    properties.sort(Comparator.comparingInt(Property::id));
    return properties;
  }
}

File: epi_judge_python/test_framework/test_failure.py
from enum import Enum, auto


class PropertyName(Enum):
    EXCEPTION_MESSAGE = auto()  # message of unhandled exception
    EXPLANATION = auto()  # explanation from TSV file
    COMMAND = auto(
    )  # last command, that caused the error, in API-testing programs
    STATE = auto(
    )  # state of the user-defined collection (for instance, in API testing)
    EXPECTED = auto()  # expected result
    RESULT = auto()  # user-produced result
    MISSING_ITEMS = auto(
    )  # list of items from input that are missing in the result set
    EXCESS_ITEMS = auto(
    )  # list of items from result that are missing in the input set
    MISMATCH_INDEX = auto(
    )  # for collections: index of the wrong item in result for binary trees: instance of TreePath describing the position of the wrong item
    EXPECTED_ITEM = auto(
    )  # value of the expected item in collection (not the whole collection)
    RESULT_ITEM = auto(
    )  # value of the result item in collection (not the whole collection)


class Property:
    def __init__(self, name, value):
        self._name = name
        self._value = value

    def name(self):
        return self._name.name.lower().replace('_', ' ')

    def id(self):
        return self._name.value

    def raw_name(self):
        return self._name

    def value(self):
        return self._value


class TestFailure(Exception):
    """
    Tested function may raise this exception
    in order to mark the current test as failed.
    """
    def __init__(self, description=''):
        super().__init__()
        self._properties = []
        self._description = description

    def with_property(self, name, value):
        self._properties.append(Property(name, value))
        return self

    def with_mismatch_info(self, mismatch_index, expected_item, result_item):
        return self\
            .with_property(PropertyName.MISMATCH_INDEX, mismatch_index)\
            .with_property(PropertyName.EXPECTED_ITEM, expected_item)\
            .with_property(PropertyName.RESULT_ITEM, result_item)

    def get_description(self):
        return self._description

    def get_max_property_name_length(self):
        return max((len(prop.name()) for prop in self._properties), default=0)

    def get_properties(self):
        self._properties.sort(key=lambda prop: prop.id())
        return self._properties
